course_catalog/api.py

```diff
@@ -52,66 +57,104 @@ def parse_mitx_json_data(course_data, force_overwrite=False):
-                max_modified, "%Y-%m-%dT%H:%M:%S.%fZ"
-            ).astimezone(pytz.utc)
-            if (
-                compare_datetime <= course_instance.last_modified
-                and not force_overwrite
-            ):
-                log.debug(
-                    "(%s, %s) skipped", course_data.get("key"), course_run.get("key")
-                )
-                continue
-            index_func = update_course
-        except Course.DoesNotExist:
-            course_instance = None
-            index_func = index_new_course
+    # Try and get the Course instance. If it exists check to see if it needs updating
+    try:
+        course = Course.objects.get(course_id=course_data.get("key"))
+        compare_datetime = datetime.strptime(
+            course_modified, "%Y-%m-%dT%H:%M:%S.%fZ"
+        ).astimezone(pytz.utc)
+        needs_update = (compare_datetime >= course.last_modified) or force_overwrite
+        index_func = update_course
+    except Course.DoesNotExist:
+        index_func = index_new_course
+        course = None
+        needs_update = True
 
-        edx_serializer = EDXSerializer(
+    if needs_update:
+        edx_serializer = EDXCourseSerializer(
             data={
-                **course_run,
+                **course_data,
                 "course_image": course_data.get("image"),
-                "max_modified": max_modified,
+                "max_modified": course_modified,
                 "raw_json": course_data,  # This is slightly cleaner than popping the extra fields inside the serializer
             },
-            instance=course_instance,
+            instance=course,
         )
 
         if not edx_serializer.is_valid():
             log.error(
                 "Course %s is not valid: %s",
-                course_run.get("key"),
+                course_data.get("key"),
                 edx_serializer.errors,
             )
-            continue
-
-        # Make changes atomically so we don't end up with partially saved/deleted data
-        with transaction.atomic():
-            course = edx_serializer.save()
+        else:
+            # Make changes atomically so we don't end up with partially saved/deleted data
+            with transaction.atomic():
+                course = edx_serializer.save()
+                # Parse each course run individually
+                for course_run in course_data.get("course_runs"):
+                    if should_skip_course(course_run.get("title")):
+                        continue
+
+                    # Get the last modified date from the course run
+                    course_run_modified = course_run.get("modified")
+
+                    # Since we use data from both course and course_run and they use different modified timestamps,
+                    # we need to find the newest changes
+                    max_modified = (
+                        course_modified
+                        if course_modified > course_run_modified
```

Comment:
These values are actually strings so we need to parse them to get the correct value.

Reply:
It can also be shortened to `max(course_modified, course_run_modified)`.

Reply:
👍